### DISCUSSION TITLE
Shared state best practices?

### DISCUSSION THREAD

**Discussion by dmeehan1968** [pointfreeco/swift-composable-architecture#879] [Q&A]
I'm exploring TCA and trying some simple applications to better understand the architecture and see how it maps to some of problems I've encountered with SwiftUI development.

I got a simple demo which implements a Parent and Child view, with corresponding TCA State, Actions and reducers.

In the demo source,

* Parent holds a list of items (Strings for simplicity)
  * is responsible for establishing navigation and searching, providing the filtered items to the Child
  * Child state is recomputed each time the query changes (real apps would likely debounce the input)
* Child view implements the List view, using ForEach on the items
  * applying the onMove and onDelete modifiers to enable delete and visual reordering, sending actions when these occur

The problem I have is that I can't figure out how to propogate changes to the filtered list back to the Parent.  I'm aware I can add a case to the Parent reducer to observe the move/delete actions occuring on the Child, but the offsets that are provided are relative to the Child's filtered version of the list, which has already been modified due to combine order (it's not clear if I can reverse the reducer order in combine without other unwanted side effects).

I've played around with a separate object type to hold the items and propogate changes but this seems undesirable as it starts to allow for state changes outside of TCA. It seems like there should be a TCA state associated with a collection of items, the 'source of truth', and a subset is passed to the Child but is linked, so that changes also occur in the Parent list.

Logically the Child view should be responsible for its own actions, i.e. operations on its own list of items should be propagated back to the parent, without the parent having to repeat the changes, or perform them on behalf of the Child so that updates trickle back to the Child. The Child view -should- both function as a standalone view, and be reusable.

I can't really fathom from the SharedState and Search examples how those techniques would apply.

Here is some boilerplate for the application, which works up to the point of needing to propagate the changes between Parent and Child.

Thanks in advance for any pointers.

<details>
  <summary>Parent.swift - Parent view, state, actions and reducer</summary>

```swift
import SwiftUI
import ComposableArchitecture

// MARK: S T A T E

struct ParentState: Equatable {
    var items: [String]
    var child: ChildState
    @BindableState var query: String = ""

    init(items: [String]) {
        self.items = items
        self.child = ChildState(items: items)
    }
}

extension ParentState {
    static var testData = [
        "First",
        "Second",
        "Third",
        "Fourth",
    ]
}

// MARK: A C T I O N

enum ParentAction: Equatable, BindableAction {
    case binding(BindingAction<ParentState>)
    case child(action: ChildAction)
}

// MARK: R E D U C E R

let parentReducer = Reducer<ParentState, ParentAction, Void>.combine(

    childReducer.pullback(
        state: \.child,
        action: /ParentAction.child,
        environment: { _ in Void() }
    ),

    Reducer { state, action, _ in

        switch action {

        case .binding(\.$query):
            state.child = ChildState(items: state.items.filter({ state.query.isEmpty || $0.lowercased().contains(state.query.lowercased()) }))

        case .child(action: .move(fromOffsets: let source, toOffset: let destination)):
            // TODO: apply to state.items, recalculate state.child to reapplying filter
            // NB: offsets are relative to child (filtered) view, so not
            // valid in parent
            break

        case .child(action: .delete(atOffsets: let source)):
            // TODO: apply to state.items, recalculate state.child to reapplying filter
            // NB: offsets are relative to child (filtered) view, so not
            // valid in parent
            break

        default:
            break
        }

        return .none
    }
    .binding()
)

// MARK: S T O R E

typealias ParentStore = Store<ParentState, ParentAction>

// MARK: V I E W

struct Parent: View {
    var store: ParentStore

    var body: some View {
        WithViewStore(self.store) { viewStore in
            NavigationView {
                IfLetStore(
                    self.store.scope(state: { $0.child }, action: ParentAction.child(action:)),
                then: { store in
                    Child(store: store)
                }) {
                    Text("No Child")
                }
                .toolbar {
                    ToolbarItem(placement: .primaryAction, content: { EditButton() })
                }
                .navigationTitle("Searchable List")
                .navigationBarTitleDisplayMode(.inline)
            }
            .navigationViewStyle(.stack)
            .searchable(text: viewStore.binding(\.$query))
        }
    }
}

// MARK: P R E V I E W

struct Parent_Previews: PreviewProvider {
    static var previews: some View {
        Parent(
            store: ParentStore(
                initialState: ParentState(
                    items: ParentState.testData
                ),
                reducer: parentReducer,
                environment: Void()
            )
        )
    }
}
```

</details>

<details>
  <summary>Child.swift - Child view, state, actions and reducer</summary>

```swift
//
//  Child.swift
//  architectures
//
//  Created by Dave Meehan on 04/11/2021.
//

import SwiftUI
import ComposableArchitecture

// MARK: S T A T E

struct ChildState: Equatable {
    var items: [String] = []
}

// MARK: A C T I O N

enum ChildAction: Equatable {
    case move(fromOffsets: IndexSet, toOffset: Int)
    case delete(atOffsets: IndexSet)
}

// MARK: R E D U C E R

let childReducer = Reducer<ChildState, ChildAction, Void> { state, action, _ in

    switch action {
    case .move(fromOffsets: let source, toOffset: let destination):
        print("Action: \(action), source: \(source.map({ $0 })), destination: \(destination)")
        // TODO: apply changes to state.items
        // NB: changes not propagated to parent, so must be repeated there
        break

    case .delete(atOffsets: let source):
        print("Action: \(action), source: \(source.map({ $0 }))")
        // TODO: apply changes to state.items
        // NB: changes not propagated to parent, so must be repeated there
        break
    }
    return .none
}

// MARK: S T O R E

typealias ChildStore = Store<ChildState, ChildAction>

// MARK: V I E W

struct Child: View {
    var store: ChildStore

    var body: some View {
        WithViewStore(self.store) { viewStore in
            if viewStore.items.count > 0 {
                List() {
                    ForEach(viewStore.items, id: \.self) { item in
                        Text(item)
                    }
                    .onMove(perform: { viewStore.send(.move(fromOffsets: $0, toOffset: $1)) })
                    .onDelete(perform: { viewStore.send(.delete(atOffsets: $0)) })
                }
            } else {
                Text("No matches")
            }
        }
    }
}

struct Child_Previews: PreviewProvider {
    static var previews: some View {
        Child(
            store: ChildStore(
                initialState: ChildState(items: ParentState.testData),
                reducer: childReducer,
                environment: Void()
            )
        )
    }
}
```

</details>

**Comment by tgrapperon** [comment]
One way to do it would be to make `child` a computed property of `Parent`:
```swift
struct ParentState: Equatable {
  var items: [String]
  @BindableState var query: String = ""

  init(items: [String]) {
    self.items = items
  }
  var child: ChildState {
    get { .init(items: items) }
    set { items = newValue.items }
  }
}
```
If your child feature have some internal state you don't want to handle with your parent, you can use a private storage property:
```swift
struct ParentState: Equatable {
  var items: [String]
  @BindableState var query: String = ""

  init(items: [String]) {
    self.items = items
  }
  private var _child: ChildState = .init(items: [])
  var child: ChildState {
    get { 
      var state = _child
      state.items = items
      return state
    }
    set { 
      _child = newValue
      items = newValue.items 
    }
  }
}
```
This allows to preserve any internal state set by the child, while ensuring it always receives an up to date `items` array from its parent.

You can furthermore append additional behaviors when reducing child actions in `parentReducer` if needed.

**Reply by dmeehan1968** [reply]
Thanks, that simplifies what's in the Shared example and makes it more digestible.

I don't believe that this copes with the requirement of the Child to receive filtered items.  Whilst operations on the Child items array would propagate back to the Parent, any filter applied would also cause those items to be removed from the Parent.  I guess it would be possible to diff the newValue items against the Parent items and figure out which items have actually been deleted, and apply those changes to the parent.

**Reply by tgrapperon** [reply]
After looking more closely at your example, and if the child feature should be allowed to work on a subset of items, I would host a filtered items list in the parent, beside the source `items` property, and pass this value to the child instead of `items`
What you do in the parent when you receive an update from the child is up to you. I think replaying the actions in the unfiltered list, and refiltering it is fine. If you want to guarantee that your filtered list is the same as the filtered items list, you can:
- Bypass the setter part in the `child` computed propery, or
- Cook a noop child reducer for these actions, and handle all the changes in the parent reducer. You can use some higher-order reducer like [`filter` from isowords](https://github.com/pointfreeco/isowords/blob/main/Sources/TcaHelpers/FilterReducer.swift) to derive such a reducer from the base child reducer. That's probably the route I would take.

**Reply by dmeehan1968** [reply]
Thanks for the additional suggestions.  I posted my own answer which has a 'two stores' solution, which might at least demonstrate a different approach to shared state.

However, your suggestions have led me to reconsidering what the views are trying to do, and I realised that the Child view doesn't need to get a filtered array - it can operate on the entire array and the `List` iteraton can then conditionally output a row if the query matches the item.

I'm actually a bit surprised here as SwiftUI seems to maintain the offsets correctly, even though there is no corresponding 'row' from the UI perspective.  I'm guessing that's because the `id` is tracked against an enumeration/iteration.  I did try it with `EmptyView()` for when the item is filtered out, but that doesn't seem to function any differently.

```swift
List() {
    ForEach(viewStore.items, id: \.self) { item in
        if viewStore.filter(item) {
            Text(item)
        }
    }
    .onMove(perform: { viewStore.send(.move(fromOffsets: $0, toOffset: $1)) })
    .onDelete(perform: { viewStore.send(.delete(atOffsets: $0)) })
}
```

This is actually vastly simpler, and uses the computed ChildState, which just makes sure its passing in a closure (`(String) -> Bool`) that's updated with the current query string

```swift
struct ParentState: Equatable {
    //...
    var child: ChildState {
        get {
            .init(items: items, filter: { query.isEmpty || $0.lowercased().contains(query.lowercased()) })
        }
        set {
            items = newValue.items
        }
    }
    // ...
}
```

The Child view can simply use the offsets given by the `onMove` and `onDelete` handlers, it operates on its own local copy, the result of any mutation being assigned back to the Parent's items property.

The only thing that seems off here is that part of the point of TCA (I beleive) is that the reducers are supposed to be the only place where state is mutated, and I worry if having computed mutating properties in the ParentState make it harder to reason about where state mutations occur.

I really appreciate the input, it's helped me to rethink the approach to the problem instead of doggedly looking for a solution to how to pass a subset list to a child view and then reconcile any mutations.

**Reply by tgrapperon** [reply]
> The only thing that seems off here is that part of the point of TCA (I beleive) is that the reducers are supposed to be the only place where state is mutated[...]

This is still the case with computed properties! If you think about it, what's triggering the setter part of the child state computed property is the pulled back child reducer working on the child substate. When you set `state.value = 2` in a child reducer, the child substate with `.value == 2` is passed up to the setter of the computed property by the `pullback` higher order reducer. All of this happen when reducing some action.
Of course, the mutation can be unaccounted explicitly by some specific, non-embedding, action in the parent reducer, but I can't see some configuration right now where any mishaps can't be spotted by some test.

State derivation is part of the model as much as action reduction: If your child state depends on an `Int` and your parent's state has two `Int` properties, the child state you derive from the parent can accept both properties as arguments. It can be something desirable (for example, you pass `.init(numberOfRows: self.isEditing ? self.editedItems.count : self.items.count)`), or it can be some programmer error (for example, you pass `.init(numberOfRows: self.colorIndex)` instead of `self.items.count`). The same degree of freedom (of making mistakes or choices) happens on the other way, on the setter side. It's ultimately something that will flow through reducers and can be tested.

Because the `State` is `inout`, the reducer signature can be rewritten from `(inout State, Action, Environment) -> Effect` as `(State, Action, Environment) -> (State, Effect)`. Excluding `Environment` and side-effects, its corresponding to `(State, Action) -> State`. In other words, the reducer takes one state and one action and returns a new state. If you consider the trivial noop reducer that does nothing, and if your state derivation (*i.e.* your computed properties logic) is incoherent, it will return incoherent states as the result of reducing actions; and this can be tested. If your `State` can host some mutating helper function like `mutating func inc() { self.count += 1 }`, you would probably have no concerns calling it when reducing some action in a reducer (`state.inc()`). Computed properties are called exactly in the same way when pulling back child domains into their parent (but you don't see them into a `switch action{…}` context): The parent reducer extracts a child stubstate from the parent state, passes-it to the child reducer, and embeds the result back into the parent. Both computed property code paths (`get` and `set`) are activated during a reduction, even if the child reducer is trivial.

**Reply by dmeehan1968** [reply]
I agree, the computed state is a side effect of the child reducer.  Thanks for the clarification.

**Reply by dmeehan1968** [reply]
I've discovered another limitation with this approach, which is that it works well as long as the computed child state is only sharing a small amount of state (i.e. the array of items in this example), but if the Child state is complex, it becomes necessary to lift all of that state into the Parent, and pass it back and forth through the computed getter/setter for the Child.  In effect, the Parent has to become responsible for persisting all aspects of the child state.  Of course, the Parent _is_ responsible for doing that, but usually does so on the Child state as a single instance, but due to immutability of the State (when a struct) it can't neatly package this up, and must hold state for each property of Child state so that the child state can be computed as necessary.

I think this is what I was doing with a 'two stores' approach - treating the Child state and the Shared state as discrete stores, and passing both as separate arguments to the Child view, making the inputs explicit.

If one wanted to take the view that the intention with TCA was for a view to receive a single state instance (which I think is implied by the examples), then to avoid exposing the Parent state to the child, an intermediate store which was a union of Child and Shared state  could be utilised.  The Child would use that intermediate store, dispatching to the individual reducers, although this makes each `viewStore.send` a little more verbose than normal.

**Reply by tgrapperon** [reply]
Please note that you can define a private storage for your child state to handle explicitly only the shared properties in the getter/setter. Let say you have:
```swift
struct Child {
  var color: Color = .red
  var dragPosition: CGPoint = .zero
}
```
Only `color` is inherited from the parent. The `dragPosition` is used to store some position managed by the child feature.

You can define your parent like
```swift
struct Parent {
  var color: Color

  private var _child: Child = .init()
  var child: Child {
    get { 
      var state = _child
      state.color = color
      return state
    }
    set { 
      _child = newValue
      color = newValue.color 
    }
  }
}
```
As you can see, you don't have to manage explicitly the `dragPosition` in the parent.

The main constraint is that you need to design your child state in a way that allows individual properties to be updated without creating an invalid substate (for example when you have some stored property which was derived from the value you're updating).

Having all the data in one struct is roughly equivalent to having it split in several structs in terms of payload for the computer. If you have difficulties keeping it in the stack, you can always box your biggest child states as reference types. You can look at #488 for more information on the subject.

**Reply by dmeehan1968** [reply]
AAh, yes, I see how this works now.  Thank you.

**Reply by dmeehan1968** [reply]
I seem to now have a problem with testing.  

When I look at the examples, this uses the `TestStore`, with `send` actions which verify the state.  This doesn't seem to play well with the private & computed child, as I can't describe in the send closure the changes that will happen (as the private child state isn't accessible - nor does it reflect the change that happens in the computed getter - and using the computed setter doesn't complete the round trip).

I've looked at the episodes on testing and the API seems to have changed.  In the examples, the reducer is called directly, with the return result appearing to be an array (assert against `isEmpty`). But what I'm now getting is a publisher.  I wouldn't want any effects to be processed, as I just need to check what they are.  I'm not familiar with the Combine API to know how I might inspect this.

In this case, I'd expect the effect returned to be `.none`.

```swift
let items = ["First", "Second", "Third"]
var state = ParentState(items: items)
let effect = remindersReducer(&state, .binding(.set(\.$query, "F")), Void())
XCTAssertEqual(state.child, ChildState(items: items, query: "F"))
// NB: How to test 'effect' to be .none
```

**Reply by tgrapperon** [reply]
The private child state storage was marked as such only for the intent. You can have it marked `internal` and you should be able to access its fields if you're using `@testable import ParentModule`. If you modularize your app with one feature per model, it is equivalent to being private for external clients.

The testing API was indeed modernized, and you can now flatten your closures and directly assert things in line when using a `TestStore` (you still conveniently build your expected state using closures). You should be able to test your parent and child using a `TestStore`, using internal storage fields or test helpers to set the values if needed. Some care may be required when architecting the state/sub-states, so they are indeed testable (i.e. buildable). You shouldn't need to assert manually with `XCTest` methods when using a `TestStore`

The `TestStore` forces you to acknowledge any effect produced by an action. It means that expecting no effect is equivalent to testing for the `.none` effect:
- If you expect some effect, and `.none` is produced, it fails;
- If you expect `.none` and, some effect is produced, it fails;

I hope it helps.

**Reply by dmeehan1968** [reply]
Thank you, yes that clarifies the use of TestStore.  I'm still unclear on whether TestStore can be passed to a view to allow for Snapshot testing, or whether the simplified TCA implemented in Episode 86 is the only way to do it.  It seems that if TestStore can be used, it would allow for only one test scenario to both validate the actions and the effect they have on the view.  Am I just barking up the wrong tree?

**Reply by tgrapperon** [reply]
I'm not using Snapshot testing, but if I did, I would use a standard TCA store.
You're ultimately not testing the same thing in both setups:
- TestStore -> You're testing the validity of your model and business logic, regardless of the UI.
- Snapshot testing -> You're testing that things (most likely the UI) stay the same as you evolve the app. This doesn't check the validity of what's represented (your reference snapshot can be wrong). In this spirit, I would put the app in the closest configuration it will effectively run on devices. `TestStore` is not a `Store`, and you won't be able to extract `ViewStore`'s. You will also need to acknowledge effects when snapshot testing, which doesn't seem to make sense.

**Reply by dmeehan1968** [reply]
Thanks, that was the clarification I needed.

**Reply by remlostime** [reply]
@tgrapperon For snapshot testing, since in `View`, we use `ViewStore` to send the action, and we cannot access `ViewStore` in our snapshot test. How do we mimic each action flow to the `ViewStore` and compare the screenshot?

**Reply by tgrapperon** [reply]
I'm not a specialist with snapshot testing, and I think @mbrandonw or @stephencelis may provide better insights, but I would create a stateless `ViewStore` (`ViewStore(store.stateless)`) to send actions to the store from outside the `View`. The `ViewStore` from the `View` should update the view accordingly. 

With snapshot testing, you want to check that the UI is reflecting your state, so I would say that, in many cases, you can even forgo the `ViewStore` and push new `Store`s with different initial states and snapshot the results.

As I'm getting it, snapshot testing is your test harness against UI regressions. It doesn't check that the submit button is visible when it should: it only checks that the snapshot is the same as the one you reviewed and validated.

For this reason, I don't think testing the business logic at the same time as snapshots (i.e., in the same test) would be a good practice if this is something you envisaged.

Now, as you will generate many states when testing the business logic, it may be possible to exploit `TestStore` to snapshot some codable state values and feed them later to a snapshot testing pass.

Again, I'm providing some very naive advice, as I'm not using snapshot testing myself (yet). Please take it with a pinch of salt.

**Reply by hibernat** [reply]
@tgrapperon I like the idea of updating the child state VERY much! Great, thanks!
I dislike initializing new child state in the getter, your idea is just brilliant.

**Comment by dmeehan1968** [comment]
I've done some more work on this and come up with a possible solution, although I'm not sure quite whether there might be some hidden gotchas.  I've not seen another example of passing more than one store to a view, but this seemed to open up some possibilities.

There were several parts to the solution:

1. Turn the array of strings into a State/Action/Reducer/Store as `ItemArray`
  1.  Each item turned into an identifable `Item`
  2. State an alias for `IdentifiableArrayOf<Item>`
  3. Use pullback to integrate it into the Parent state.
2. Move the `@BindableState var query: String` property into the Child state
3. Pass both the Child store and the Item store into the Child view
4. Send the CRUD actions to the Item Store from within the child view
5. Sprinkle some syntactic sugar via extensions to the ItemArray to aid with the conversion from IndexSet/Int to Item.ID
6. I opted to group the State/Action/Reducer/Store under a struct that can't be instantiated (`private init() { }`) as a way of namespacing (I know people do with `enum` but the reducer can't be placed in an enum, so this seems to be an alternative).

<details>
<summary>Item.swift</summary>

```swift
import Foundation

struct Item: Identifiable, Equatable {
    var id = UUID()
    var title: String
}
```
</details>

<details>
<summary>ItemArray.swift</summary>

```swift
import Foundation
import ComposableArchitecture

struct ItemArray {

    private init() {}

    // MARK: S T A T E

    typealias State = IdentifiedArrayOf<Item>

    // MARK: A C T I O N

    enum Action: Equatable {
        case append(Item)
        case update(Item)
        case move(ids: [Item.ID], toId: Item.ID?)
        case remove(ids: [Item.ID])
    }

    // MARK: R E D U C E R

    static let reducer = Reducer<State, Action, Void> { state, action, _ in
        switch action {

        case .append(let item):
            state.append(item)

        case .update(let item):
            guard let index = state.firstIndex(where: { $0.id == item.id }) else { break }
            state.update(item, at: index)

        case .move(ids: let ids, toId: let toId):
            // convert Item.ID into [Int] offsets
            let source = ids
                // convert Id to offset?
                .map({ id in state.firstIndex(where: { $0.id == id })})
                // remove nil
                .filter({ $0 != nil })
                // unwrap
                .map({ $0! })

            // dest is either valid offset or out of range
            let dest = state.firstIndex(where: { $0.id == toId })
            // move source offsets to dest offset (or end of list if nil)
            state.move(fromOffsets: IndexSet(source), toOffset: dest ?? state.count)

        case .remove(ids: let ids):
            // convert Item.ID to [Int] offsets
            let source = ids.map({ id in state.firstIndex(where: { $0.id == id })}).filter({ $0 != nil }).map({ $0! })
            state.remove(atOffsets: IndexSet(source))

        }
        return .none
    }

    // MARK: S T O R E

    typealias Store = ComposableArchitecture.Store<State, Action>

}
```

</details>

<details>
<summary>Parent.swift</summary>

```swift
import SwiftUI
import ComposableArchitecture

struct Parent {

    // MARK: S T A T E

    struct State: Equatable {
        var items: ItemArray.State
        var child: Child.State

        init(items: [Item]) {
            self.items = ItemArray.State(uniqueElements: items)
            self.child = Child.State()
        }
    }

    // MARK: T E S T   D A T A

    static var testData = [
        Item(title: "Item 1"),
        Item(title: "Item 2"),
        Item(title: "Item 3"),
        Item(title: "Item 4"),
    ]

    // MARK: A C T I O N

    enum Action: Equatable, BindableAction {
        case binding(BindingAction<State>)
        case child(action: Child.Action)
        case items(action: ItemArray.Action)
    }

    // MARK: R E D U C E R

    static let reducer = Reducer<State, Action, Void>.combine(

        ItemArray.reducer.pullback(
            state: \.items,
            action: /Action.items,
            environment: { _ in Void() }
        ),

        Child.reducer.pullback(
            state: \.child,
            action: /Action.child,
            environment: { _ in Void() }
        ),

        Reducer { state, action, _ in
            return .none
        }
        .binding()
    )

    // MARK: S T O R E

    typealias Store = ComposableArchitecture.Store<State, Action>

    // MARK: V I E W

    struct View: SwiftUI.View {
        var store: Store

        var body: some SwiftUI.View {
            WithViewStore(self.store) { viewStore in
                NavigationView {
                    Child.View(
                        childStore: self.store.scope(state: { $0.child }, action: Parent.Action.child(action:)),
                        itemStore: self.store.scope(state: { $0.items }, action: Parent.Action.items(action:))
                    )
                    .toolbar {
                        ToolbarItem(placement: .primaryAction, content: { EditButton() })
                        ToolbarItem(placement: .bottomBar) {
                            Button("Add Item") { viewStore.send(.items(action:.append(Item(title: "Item \(viewStore.items.count+1)"))))}
                        }
                    }
                    .navigationTitle("Searchable List")
                    .navigationBarTitleDisplayMode(.inline)
                }
                .navigationViewStyle(.stack)
                .searchable(text: viewStore.binding(\.child.$query))
            }
        }
    }

}

// MARK: P R E V I E W

struct Parent_Previews: PreviewProvider {
    static var previews: some View {
        Parent.View(
            store: Parent.Store(
                initialState: Parent.State(
                    items: Parent.testData
                ),
                reducer: Parent.reducer,
                environment: Void()
            )
        )
    }
}
```

</details>

<details>
<summary>Child.swift</summary>

```swift
import SwiftUI
import ComposableArchitecture

struct Child {

    private init() {}

    // MARK: S T A T E

    struct State: Equatable {
        @BindableState var query: String = ""
    }

    // MARK: A C T I O N

    enum Action: Equatable, BindableAction {
        case binding(BindingAction<State>)
    }

    // MARK: R E D U C E R

    static let reducer = Reducer<State, Action, Void> { state, action, _ in

        return .none
    }
    .binding()

    // MARK: S T O R E

    typealias Store = ComposableArchitecture.Store<State, Action>

    // MARK: V I E W

    struct View: SwiftUI.View {
        var childStore: Store
        var itemStore: ItemArray.Store

        var body: some SwiftUI.View {
            WithViewStore(self.childStore) { childViewStore in
                WithViewStore(self.itemStore) { itemViewStore in
                    let filtered = itemViewStore.state.filtered(query: childViewStore.query)
                    if filtered.count > 0 {
                        List {
                            ForEach(filtered) { item in
                                HStack {
                                    Text(item.title)
                                    Spacer()
                                    Button("Update") {
                                            itemViewStore.send(.update(Item(id: item.id, title: item.title + ".")))
                                        }
                                }
                            }
                            .onMove { itemViewStore.onMove(items: filtered, fromOffsets: $0, toOffset: $1) }
                            .onDelete { itemViewStore.onDelete(items: filtered, atOffsets: $0) }
                        }
                    } else {
                        Text("No matches")
                    }
                }
            }
        }
    }
}

// MARK: E X T E N S I O N

extension ItemArray.State where Element == Item {
    func filtered(query: String) -> IdentifiedArrayOf<Element> {
        self.filter({
            query.isEmpty || $0.title.lowercased().contains(query.lowercased())
        })
    }

    func mapIds(from indexset: IndexSet) -> [Element.ID] {
        indexset.map({ self[$0].id })
    }

    subscript(safe index: Int) -> Element.ID? {
        self.indices.contains(index) ? self[index].id : nil
    }
}

extension ViewStore where State == ItemArray.State, Action == ItemArray.Action {
    func onMove(items: ItemArray.State, fromOffsets: IndexSet, toOffset: Int) -> Void {
        self.send(
            .move(
                ids: items.mapIds(from: fromOffsets),
                toId: items[safe: toOffset]
            )
        )
    }

    func onDelete(items: ItemArray.State, atOffsets: IndexSet) -> Void {
        self.send(
            .remove(
                ids: items.mapIds(from: atOffsets)
            )
        )
    }
}

struct Child_Previews: PreviewProvider {

    static var previews: some View {
        Child.View(
            childStore: Child.Store(
                initialState: Child.State(),
                reducer: Child.reducer,
                environment: Void()
            ),
            itemStore: ItemArray.Store(
                initialState: ItemArray.State(uniqueElements: Parent.testData),
                reducer: ItemArray.reducer,
                environment: Void()
            )
        )
    }
}

```

</details>

**Comment by mfclarke** [comment]
I'm still quite new to Composable Architecture, so please bear with me. Why isn't this a simple case of the parent reducer using the items array in the child property of state to translate the index to the full array and delete/move in the parent items accordingly?

This seems like a pretty basic use case and I'm surprised the answer would be to use 2 stores?

**Reply by mfclarke** [reply]
Ah sorry, the filtered array has already been modified by the time the parent reducer gets the actions. What about the parent reducer holding onto the previous filtered array in it's state as well? And updating that to the new filtered whenever it receives child move/delete actions, after applying the move/delete referencing the old filtered?

**Reply by dmeehan1968** [reply]
Whether the child collection has been modified or not is dependent on the order the reducers are combined.  In practive I've realised that its less likely that you'd use index based logic, and more likely use id based logic (so the child translates the indexes provided by ForEach's onMove/onDelete into id's, then sends those as an action to the reducer - the parent can then intercept the child action and make the same changes to its own unfiltered collection.  The simplicity of my sample actually made it hard to solve that adding a little complexity elsewhere.

**Reply by mfclarke** [reply]
Ah yes that all makes sense, and it's a much simpler solution than having 2 stores. Nice.

I feel like this should be one of the case studies. This scenario would crop up in a lot of apps. Even though in this particular example I would personally put it all into 1 reducer (since it's all the same screen/feature anyway), there are definitely cases where state needs to be shared across features where it's represented and worked on differently in each feature.
